### PR TITLE
Issue #SB-2176 feat: Signup module

### PR DESCRIPTION
**Pull request by vaishnavimanjunath**

### CODE REVIEW

File: src/app/client/src/app/modules/core/components/main-header/main-header.component.html [hunk 1]
```diff
@@ -20,21 +21,23 @@
...
+                    <a class="item" href="/logoff">{{resourceService.frmelmnts?.lnk?.logout}}</a>
 
                 </div>
             </div>
         </div>
         <ng-template #showLoginBtn>
             <div class="right floated one wide column tablet computer only">
                 <div class="beforelogin ui secondary pointing menu padding-top-1rem">
-                    <button class="ui item">{{resourceService.frmelmnts?.btn?.signup}}</button>
+                    <!-- <a> -->
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/core/components/main-header/main-header.component.html]: Remove commented code.

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.html [hunk 2]
```diff
@@ -3,8 +3,8 @@
   <div class="header announcementCard-header">
     <span class="header cardsHeading">{{resourceService?.frmelmnts?.lbl?.certificationAward}}</span>
   </div>
-  <div *ngIf="badgeArray && badgeArray.length" class="ui items margin-top-10">
-    <div class="item" *ngFor="let badge of badgeArray">
+  <div *ngIf="badgeArray" class="ui items margin-top-10">
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.html]: Not as per latest change in angular-migration branch.

Expected:
<div *ngIf="badgeArray && badgeArray.length" class="ui items margin-top-10">

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.html [hunk 3]
```diff
@@ -17,7 +17,23 @@
           </div>
         </div>
       </div>
-      <div *ngIf="badge.createdTS || badge.createdTs" class="meta">{{badge.createdTS || badge.createdTs | dateFormat}}</div>
+      <div class="meta">{{badge.createdDate || badge.createdTs | dateFormat: 'MMM YYYY'}}</div>
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.html]: Fix not applied on latest change in angular-migration branch.

Expected:
<div *ngIf="badge.createdTS || badge.createdTs" class="meta">{{badge.createdTS || badge.createdTs | dateFormat: 'MMM YYYY'}}</div>

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.html [hunk 4]
```diff
@@ -17,7 +17,23 @@
...
+    </div>
+    <div class="ui bottom attached tabular menu userSkillMore ng-scope " *ngIf="badgeArray.length > defaultLimit">
+      <div class="active item addSkillMoreBtn">
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.html]: Rename CSS class addSkillMoreBtn to addViewMoreBtn.

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.ts [hunk 5]
```diff
@@ -13,6 +13,10 @@ export class ProfileBadgeComponent implements OnInit {
  * Reference of User Profile interface
  */
   userProfile: IUserProfile;
+  viewMore = true;
+  defaultLimit = 3;
+  limit = 3; // config
+  badgeData: any;
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.ts]: Not as per latest change in angular-migration branch.

Expected:
Unused variable badgeData is removed.

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts [hunk 6]
```diff
@@ -27,12 +27,29 @@ describe('ProfileBadgeComponent', () => {
     // fixture.detectChanges();
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts]: Remove commented code.

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.ts [hunk 7]
```diff
@@ -48,4 +52,13 @@ export class ProfileBadgeComponent implements OnInit {
         }
       });
   }
+  toggle(lim) {
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.ts]: Rename lim to isViewLess or isViewMore (requires changing toggle condition).

File: src/app/client/src/app/modules/public/components/signup/signup.component.ts [hunk 8]
```diff
@@ -1,13 +1,16 @@
...
+import { Router } from '@angular/router';
 
 @Component({
   selector: 'app-signup',
   templateUrl: './signup.component.html',
   styleUrls: ['./signup.component.css']
 })
 export class SignupComponent implements OnInit {
+  @ViewChild('modal') modal;
   /**
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/public/components/signup/signup.component.ts]: Code is not formatted in this file.

File: src/app/client/src/app/modules/public/components/signup/signup.component.ts [hunk 9]
```diff
@@ -16,27 +19,55 @@ export class SignupComponent implements OnInit {
 * Contains reference of FormBuilder
 */
   sbFormBuilder: FormBuilder;
+  /**
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/public/components/signup/signup.component.ts]: Code is not formatted in entire file.

File: src/app/client/src/app/modules/public/components/signup/signup.component.ts [hunk 10]
```diff
@@ -16,27 +19,55 @@ export class SignupComponent implements OnInit {
...
+      Validators.pattern(/^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[a-z]{2,4}$/)]),
+      language: new FormControl(null, [Validators.required])
+    });
+    this.signUpForm.valueChanges.map((value) => {
+      if (value.firstName !== null && value.userName !== null) {
+        value.firstName = value.firstName.trim();
+        value.userName = value.userName.trim();
+        return value;
+      }
+    }).subscribe((value) => {
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/public/components/signup/signup.component.ts]: Why subscribe when nothing is being done?

File: src/app/client/src/app/modules/public/services/signup.service.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  private formatRequest(request) {
+    return {
+      params: {},
+      request: request
+    };
+  }
+  signup(req) {
+    const data = this.formatRequest(req);
+    const options = {
+      url: this.configService.urlConFig.URLS.USER.SIGNUP,
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/public/services/signup.service.ts]: Rename urlConFig to urlConfig.

File: src/app/client/src/app/modules/search/components/user-profile/user-profile.component.html [hunk 12]
```diff
@@ -188,7 +188,7 @@ <h5 class="ui right floated basic icon circular button" [queryParams]="queryPara
                   </div>
                 </div>
               </div>
-              <div *ngIf="badge.createdTS || badge.createdTs" class="meta">{{badge.createdTS || badge.createdTs | dateFormat}}</div>
+              <div class="meta">{{badge.createdDate || badge.createdTs | dateFormat: 'MMM YYYY'}}</div>
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/search/components/user-profile/user-profile.component.html]: Not as per latest change in angular-migration branch.

Expected:
< div *ngIf="badge.createdTS || badge.createdTs" class="meta" >{{badge.createdTS || badge.createdTs | dateFormat}}< /div >

File: src/app/client/src/app/modules/public/components/signup/signup.component.html [hunk 13]
```diff
@@ -1,34 +1,34 @@
...
-          <div class="required field ">
+          <div class="required field" [ngClass]="signUpForm.controls.userName.invalid && signUpForm.controls['userName'].errors.required && !signUpForm.controls['userName'].pristine || (signUpForm.controls['userName'].errors && signUpForm.controls['userName'].errors.pattern) ? 'error' : ''">
             <label>{{resourceService?.frmelmnts?.lbl?.username}}</label>
             <input type="text" formControlName="userName" placeholder="eg:user.name , min 5 character required">
+            <div class="margin-bottom-10 ui pointing red basic label below" *ngIf="signUpForm.controls.userName.invalid && signUpForm.controls['userName'].errors.required && !signUpForm.controls['userName'].pristine || (signUpForm.controls['userName'].errors && signUpForm.controls['userName'].errors.pattern)">
+              {{resourceService?.messages?.stmsg?.m0087}}
+            </div>
           </div>
-          <div class="required field">
+          <div class="required field" [ngClass]="signUpForm.controls.password.invalid && signUpForm.controls['password'].errors.required && !signUpForm.controls['password'].pristine ? 'error' : ''">
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/public/components/signup/signup.component.html]: Does this cover the error when password is consisting of only whitespaces (one or more)?

File: src/app/client/src/app/modules/public/components/signup/signup.component.html [hunk 14]
```diff
@@ -38,32 +38,41 @@ <h5 class="ui basic icon circular button deny">
...
     </form>
-    <div>
-      <!--  ng-if="newUser.loader.showLoader" -->
-
-      <!-- <app-loader data="{{newUser.loader}}"></app-loader> -->
+    <div *ngIf="showLoader">
+      <app-loader></app-loader>
     </div>
     <div>
       <!--  ng-if="newUser.error.showError" -->
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/public/components/signup/signup.component.html]: Remove commented code from markup.

File: src/app/client/src/app/modules/public/components/landing-page/landing-page.component.spec.ts [hunk 15]
```diff
@@ -1,14 +1,15 @@
...
+import { NO_ERRORS_SCHEMA } from '@angular/core';
 
 describe('LandingPageComponent', () => {
   let component: LandingPageComponent;
   let fixture: ComponentFixture<LandingPageComponent>;
 
   beforeEach(async(() => {
     TestBed.configureTestingModule({
-      declarations: [ LandingPageComponent ]
+      declarations: [ LandingPageComponent ],
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/public/components/landing-page/landing-page.component.spec.ts]: Check formatting.

File: src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts [hunk 16]
```diff
@@ -5,19 +5,29 @@ import { SuiModule } from 'ng2-semantic-ui';
...
-      imports: [SuiModule, SharedModule, HttpClientTestingModule],
-      providers: [FormBuilder, ResourceService],
+      declarations: [SignupComponent],
+      imports: [SuiModule, SharedModule, CoreModule, HttpClientTestingModule],
+      providers: [FormBuilder, ResourceService, SignupService, LearnerService,
+        { provide: Router, useClass: RouterStub }],
       schemas: [NO_ERRORS_SCHEMA]
     })
-    .compileComponents();
+      .compileComponents();
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts]: Check formatting.

File: src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts [hunk 17]
```diff
@@ -26,7 +36,26 @@ describe('SignupComponent', () => {
...
+    const router = TestBed.get(Router);
+    component.redirect();
+    expect(router.navigate).toHaveBeenCalledWith(['']);
+  });
+  it('should call onSubmitForm method and return success response', () => {
+    const router = TestBed.get(Router);
+    const resourceService = TestBed.get(ResourceService);
+    resourceService.messages = mockSignUpResponse.resourceBundle.messages;
+    const signupService = TestBed.get(SignupService);
+    spyOn(signupService, 'signup').and.callFake(() => Observable.of(mockSignUpResponse.successRespone));
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts]: Rename successRespone to successResponse.

File: src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts [hunk 18]
```diff
@@ -26,7 +36,26 @@ describe('SignupComponent', () => {
...
+    component.onSubmitForm();
+    expect(router.navigate).toHaveBeenCalledWith(['']);
+  });
+  it('should call onSubmitForm method and return failure response', () => {
+    const router = TestBed.get(Router);
+    const resourceService = TestBed.get(ResourceService);
+    resourceService.messages = mockSignUpResponse.resourceBundle.messages;
+    const signupService = TestBed.get(SignupService);
+    spyOn(signupService, 'signup').and.callFake(() => Observable.of(mockSignUpResponse.failureResponse));
+    component.onSubmitForm();
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts]: expect statements are missing for this test case.

File: src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts [hunk 19]
```diff
@@ -26,7 +36,26 @@ describe('SignupComponent', () => {
     fixture.detectChanges();
   });
 
-  it('should create', () => {
-    expect(component).toBeTruthy();
+  it('should call redirect method', () => {
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts]: Add several test cases for invalid user input for each form field.

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.ts [hunk 20]
```diff
@@ -13,6 +13,10 @@ export class ProfileBadgeComponent implements OnInit {
  * Reference of User Profile interface
  */
   userProfile: IUserProfile;
+  viewMore = true;
+  defaultLimit = 3;
+  limit = 3; // config
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.ts]: Set limit to defaultLimit.

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.ts [hunk 21]
```diff
@@ -48,4 +52,13 @@ export class ProfileBadgeComponent implements OnInit {
         }
       });
   }
+  toggle(lim) {
+    if (lim === true) {
+      this.limit = this.badgeArray.length;
+      this.viewMore = false;
+    } else {
+      this.viewMore = true;
+      this.limit = 3;
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.ts]: Set limit to defaultLimit.

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.ts [hunk 22]
```diff
@@ -13,6 +13,10 @@ export class ProfileBadgeComponent implements OnInit {
  * Reference of User Profile interface
  */
   userProfile: IUserProfile;
+  viewMore = true;
+  defaultLimit = 3;
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.ts]: Shouldn't this come from a configuration file?

File: src/app/client/src/app/modules/public/services/signup.service.spec.ts [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+      'firstName': 'vaish',
+      'language': ['English'],
+      'lastName': '',
+      'password': 'password',
+      'phone': '7899918817',
+      'userName': 'vaish'
+    };
+    spyOn(signupService, 'formatRequest').and.callThrough();
+    signupService.formatRequest(req);
+    expect(signupService.formatRequest).toHaveBeenCalled();
```
Comment (bvinayakumar) [inline src/app/client/src/app/modules/public/services/signup.service.spec.ts]: Rather we should check the return value.

File: src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts (hunk 19 above)
Comment (vaishnavimanjunath) [reply src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts]: I've added

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts [hunk 24]
```diff
@@ -24,15 +24,31 @@ describe('ProfileBadgeComponent', () => {
...
     fixture = TestBed.createComponent(ProfileBadgeComponent);
     component = fixture.componentInstance;
-    // fixture.detectChanges();
   });
 
-  it('should create', () => {
+  it('should call getBadgeData method', () => {
     const userService = TestBed.get(UserService);
     const badgeService = TestBed.get(BadgesService);
     userService._userProfile = mockRes.data.userProfile;
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts]: remove

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts [hunk 25]
```diff
@@ -24,15 +24,31 @@ describe('ProfileBadgeComponent', () => {
...
     userService._userData$.next({ err: null, userProfile: mockRes.data.userProfile });
-    spyOn(badgeService, 'getAllBadgeList').and.callFake(() => Observable.of(mockRes.badgeList));
+    spyOn(badgeService, 'getDetailedBadgeAssertions').and.callFake(() => Observable.of(mockRes.badgeList));
+    spyOn(component, 'ngOnInit').and.callThrough();
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts]: why are we spying on ngOnit

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts [hunk 26]
```diff
@@ -24,15 +24,31 @@ describe('ProfileBadgeComponent', () => {
...
     component.ngOnInit();
+    expect(component.ngOnInit).toHaveBeenCalled();
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts]: why we need this

Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts]: add the assertion to badge storing property

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts [hunk 27]
```diff
@@ -24,15 +24,31 @@ describe('ProfileBadgeComponent', () => {
...
+  });
+  it('should call toggle method with limit greater than 3', () => {
+    const badgeService = TestBed.get(BadgesService);
+    const limit = true;
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts]: why wee need this

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts [hunk 28]
```diff
@@ -24,15 +24,31 @@ describe('ProfileBadgeComponent', () => {
...
+    spyOn(badgeService, 'getDetailedBadgeAssertions').and.callFake(() => Observable.of(mockRes.badgeList));
+    component.badgeArray = [];
+    component.badgeArray.length = 5;
+    component.limit = component.badgeArray.length;
+    component.toggle(limit);
+    expect(component.viewMore).toBe(false);
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts]: add more assertion

File: src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts [hunk 29]
```diff
@@ -24,15 +24,31 @@ describe('ProfileBadgeComponent', () => {
...
+    spyOn(component, 'ngOnInit').and.callThrough();
     component.ngOnInit();
+    expect(component.ngOnInit).toHaveBeenCalled();
+  });
+  it('should call toggle method with limit greater than 3', () => {
+    const badgeService = TestBed.get(BadgesService);
+    const limit = true;
+    spyOn(badgeService, 'getDetailedBadgeAssertions').and.callFake(() => Observable.of(mockRes.badgeList));
+    component.badgeArray = [];
+    component.badgeArray.length = 5;
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/profile-badge/profile-badge.component.spec.ts]: what is this operation

File: src/app/client/src/app/modules/profile/components/profile-page/profile-page.component.ts [hunk 30]
```diff
@@ -9,18 +9,32 @@ import * as $ from 'jquery';
...
+   * Contains loader message to display
+   */
   loaderMessage = {
     headerMessage: '',
     loaderMessage: 'Loading profile ...'
   };
+  /**
+   * Contains list of contributions
+   */
   contributions: any;
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/profile-page/profile-page.component.ts]: interface

File: src/app/client/src/app/modules/profile/components/profile-visibility/profile-visibility.component.ts [hunk 31]
```diff
@@ -45,13 +67,10 @@ export class ProfileVisibilityComponent implements OnInit {
     this.profileService.updateProfileFieldVisibility(req).subscribe(res => {
       this.loader = false;
       this.visibility = value;
-      // toaster suc
       this.toasterService.success(this.resourceService.messages.smsg.m0040);
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/profile-visibility/profile-visibility.component.ts]: show error message from api response param, if not there then from resource

File: src/app/client/src/app/modules/profile/components/user-additional-info/edit-user-additional-info/edit-user-additional-info.component.ts [hunk 32]
```diff
@@ -11,17 +11,44 @@ import { ProfileService } from '../../../services/profile/profile.service';
...
   /**
-* Reference of User Profile interface
-*/
+  * Reference of User Profile interface
+  */
   userProfile: IUserProfile;
+  /**
+  * Reference of Input annotation
+  */
   @Input() basicInfo: any;
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/user-additional-info/edit-user-additional-info/edit-user-additional-info.component.ts]: add interface

File: src/app/client/src/app/modules/profile/components/user-additional-info/edit-user-additional-info/edit-user-additional-info.component.ts [hunk 33]
```diff
@@ -11,17 +11,44 @@ import { ProfileService } from '../../../services/profile/profile.service';
...
+  /**
+  * Contains array of subjects which comes from config
+  */
   subjects: any;
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/user-additional-info/edit-user-additional-info/edit-user-additional-info.component.ts]: add type

File: src/app/client/src/app/modules/profile/components/user-additional-info/edit-user-additional-info/edit-user-additional-info.component.ts [hunk 34]
```diff
@@ -11,17 +11,44 @@ import { ProfileService } from '../../../services/profile/profile.service';
...
+  /**
+  * Contains Date object instance
+  */
   initDate = new Date();
+  /**
+  * Conatins array of languages which comes from config
+  */
   languages: any;
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/user-additional-info/edit-user-additional-info/edit-user-additional-info.component.ts]: add type for all

File: src/app/client/src/app/modules/profile/components/user-additional-info/user-additional-info/user-additional-info.component.ts [hunk 35]
```diff
@@ -18,29 +18,29 @@ import * as moment from 'moment';
...
- */
+   * Reference of User Profile interface
+   */
   userProfile: IUserProfile;
   /**
- * Contains social media links
- */
+   * Contains social media links
+   */
   webPages: any = {};
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/user-additional-info/user-additional-info/user-additional-info.component.ts]: add type

File: src/app/client/src/app/modules/profile/components/user-address/user-address/user-address.component.html [hunk 36]
```diff
@@ -39,7 +39,7 @@
                 </div>
                 <div *ngIf="showError || showAddressError" class="ui warning message">
                     <i class="close icon "></i>
-                    <p>You selected both address types are same, please change any one.</p>
+                    <p>Multiple addresses with same type selected. Please change any one.</p>
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/user-address/user-address/user-address.component.html]: should come from resource service

File: src/app/client/src/app/modules/profile/components/user-skills/user-skills/user-skills.component.spec.ts [hunk 37]
```diff
@@ -225,8 +40,8 @@ describe('UserSkillsComponent', () => {
   it('should create', () => {
     const userService = TestBed.get(UserService);
     const activatedRoute = TestBed.get(ActivatedRoute);
-    userService._userProfile = data.userProfile;
-    userService._userData$.next({ err: null, userProfile: data.userProfile });
+    userService._userProfile = mockSkillResponse.userProfile;
```
Comment (anoophm) [inline src/app/client/src/app/modules/profile/components/user-skills/user-skills/user-skills.component.spec.ts]: remove

File: src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts [hunk 38]
```diff
@@ -2,31 +2,128 @@ import { async, ComponentFixture, TestBed } from '@angular/core/testing';
...
+  }
+
   beforeEach(async(() => {
     TestBed.configureTestingModule({
-      declarations: [ SignupComponent ],
-      imports: [SuiModule, SharedModule, HttpClientTestingModule],
-      providers: [FormBuilder, ResourceService],
+      declarations: [SignupComponent],
+      imports: [FormsModule, ReactiveFormsModule, SuiModule, SharedModule, CoreModule, HttpClientTestingModule],
+      providers: [FormBuilder, ResourceService, SignupService, LearnerService,
```
Comment (anoophm) [inline src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts]: remove all providers

File: src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts [hunk 39]
```diff
@@ -2,31 +2,128 @@ import { async, ComponentFixture, TestBed } from '@angular/core/testing';
...
   });
 
-  it('should create', () => {
-    expect(component).toBeTruthy();
+  it('should call redirect method', () => {
+    const router = TestBed.get(Router);
+    component.redirect();
+    expect(router.navigate).toHaveBeenCalledWith(['']);
+  });
+  it('should show validation error message for form', () => {
```
Comment (anoophm) [inline src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts]: whats the purpose of this

File: src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts [hunk 40]
```diff
@@ -2,31 +2,128 @@ import { async, ComponentFixture, TestBed } from '@angular/core/testing';
...
+    component.ngOnInit();
+    expect(component.signUpForm.valid).toBeFalsy();
+  });
+  it('should show validation error message for userName', () => {
+    component.ngOnInit();
+    let errors = {};
```
Comment (anoophm) [inline src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts]: not needed

File: src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts [hunk 41]
```diff
@@ -2,31 +2,128 @@ import { async, ComponentFixture, TestBed } from '@angular/core/testing';
...
+    const userName = component.signUpForm.controls['userName'];
+    errors = userName.errors || {};
+    expect(errors['required']).toBeTruthy();
```
Comment (anoophm) [inline src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts]: just check  userName.errors. required

File: src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts (hunk 39 above)
Comment (anoophm) [inline src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts]: all this spec should be tested after changing input and should check error message as well. Discusses this with me

File: src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts [hunk 42]
```diff
@@ -2,31 +2,128 @@ import { async, ComponentFixture, TestBed } from '@angular/core/testing';
...
+  });
+  it('should call onSubmitForm method and return success response', () => {
+    const router = TestBed.get(Router);
+    component.signUpForm = new FormGroup({});
+    const resourceService = TestBed.get(ResourceService);
+    resourceService.messages = mockSignUpResponse.resourceBundle.messages;
+    const signupService = TestBed.get(SignupService);
+    spyOn(signupService, 'signup').and.callFake(() => Observable.of(mockSignUpResponse.successResponse));
+    component.onSubmitForm();
+    expect(router.navigate).toHaveBeenCalledWith(['']);
```
Comment (anoophm) [inline src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts]: should redirect to home

File: src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts [hunk 43]
```diff
@@ -2,31 +2,128 @@ import { async, ComponentFixture, TestBed } from '@angular/core/testing';
...
+    expect(mockSignUpResponse.resourceBundle.messages.smsg.m0039).toBe('Sign up successfully, Please login...');
```
Comment (anoophm) [inline src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts]: what is this assertion, remove this. Check toaster service tohave been called with that message.

File: src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts [hunk 44]
```diff
@@ -2,31 +2,128 @@ import { async, ComponentFixture, TestBed } from '@angular/core/testing';
...
+  });
+  it('should call onSubmitForm method and return failure response', () => {
+    const router = TestBed.get(Router);
+    component.signUpForm = new FormGroup({});
+    const resourceService = TestBed.get(ResourceService);
+    resourceService.messages = mockSignUpResponse.resourceBundle.messages;
+    const signupService = TestBed.get(SignupService);
+    spyOn(signupService, 'signup').and.callFake(() => Observable.of(mockSignUpResponse.failureResponse));
+    component.onSubmitForm();
```
Comment (anoophm) [inline src/app/client/src/app/modules/public/components/signup/signup.component.spec.ts]: vinaya has mentioned, still assertion are missing

File: src/app/client/src/app/modules/public/components/signup/signup.component.ts [hunk 45]
```diff
@@ -1,42 +1,80 @@
...
-    this.signUpForm = this.sbFormBuilder.group({
-      userName: ['', null],
-      password: ['', null],
-      firstName: ['', null],
-      lastName: ['', null],
-      phone: ['', null],
-      email: ['', null],
-      language: ['', null]
+    this.signUpForm = new FormGroup({
+      userName: new FormControl(null, [Validators.required, Validators.pattern('^[-\\w\.\\$@\*\\!]{5,256}$')]),
```
Comment (anoophm) [inline src/app/client/src/app/modules/public/components/signup/signup.component.ts]: move all patterns to config

File: src/app/client/src/app/modules/public/components/signup/signup.component.ts [hunk 46]
```diff
@@ -1,42 +1,80 @@
...
   /**
-* Contains reference of FormBuilder
-*/
+  * Contains reference of FormBuilder
+  */
   sbFormBuilder: FormBuilder;
+  /**
+  * Contains list of languages from config file
+  */
+  languages: any;
```
Comment (anoophm) [inline src/app/client/src/app/modules/public/components/signup/signup.component.ts]: any should not be given

**Pull request merged**